test/oauth_test.go

```diff
@@ -233,6 +234,192 @@ var _ = Describe("OAuth Strategy", func() {
+                                })
+                        })
+                })
+
+                Describe("When the service tries to update the token in an invalid way", func() {
+                        Context("When the token is updated with an invalid token value after successfully connecting", func() {
+                                It("should fail to reconnect", func() {
+                                        var err error
+
+                                        messagingService, err = builder.WithAuthenticationStrategy(config.OAuth2Authentication(
+                                                tokenC,
+                                                tokenB,
+                                                "",
+                                        )).WithReconnectionRetryStrategy(config.RetryStrategyParameterizedRetry(1, 200*time.Millisecond)).Build()
+                                        Expect(err).ToNot(HaveOccurred())
+
+                                        helpers.ConnectMessagingService(messagingService)
+
+                                        // We are passing a non-empty string as the value for a valid token property, so we expect the update
+                                        // to not return any errors. Instead the error is expected to be returned later when we try to
+                                        // reconnect using the invalid token.
+                                        err = messagingService.UpdateProperty(config.AuthenticationPropertySchemeOAuth2OIDCIDToken, "invalid token")
+                                        Expect(err).ToNot(HaveOccurred())
+
+                                        err = messagingService.UpdateProperty(config.AuthenticationPropertySchemeOAuth2AccessToken, "invalid token")
+                                        Expect(err).ToNot(HaveOccurred())
+
+                                        reconnectChan := make(chan struct{})
+                                        messagingService.AddReconnectionListener(func(event solace.ServiceEvent) {
+                                                fmt.Println("Service reconnect found")
+                                                close(reconnectChan)
+                                        })
+
+                                        helpers.ForceDisconnectViaSEMPv2(messagingService)
+                                        Consistently(reconnectChan).ShouldNot(Receive())
+                                        Eventually(messagingService.IsConnected()).Should(BeFalse())
+
+                                        // The service should fail to reconnect above, so if the service is connected at this point,
+                                        // then there was an error that was not detected, so we will fail the test here, after
+                                        // cleaning up the service.
+                                        if messagingService.IsConnected() {
```

Comment:
We should disconnect the messaging service regardless after this point. This way we can verify the context is destroyed in case for some reason it did not terminate (e.g. if we accidentally set the retry strategy to forever instead of once). This will prevent leaks. This test is fine as is since we do set the reconnect retry strategy, just worth considering.

Reply:
I reworked the order of operations so that if the service is still connected at the end of the test, the test will fail, and the service is disconnected at the end of the test, before succeeding or failing, just in case there is undefined behaviour resulting in the service still being connected by the end of the test.